feat(lazy): Add memoized LazyValue

ndench opened the pull request:
As discussed, this PR introduces the ability to memoize a LazyValue, which enables the following use-case:

```php
class TaskFactory extends ModelFactory
{
    protected function getDefaults(): array
    {
        $owner = LazyValue::memoize(fn () => UserFactory::new());

        return [
            // Same User instance will be both added to the Project and set as the Task owner
            'project' => ProjectFactory::new(['users' => [$owner]]),
            'owner'   => $owner,
        ];
    }
    
    protected static function getClass(): string
    {
        return Task::class;
    }
}
```

I'm not entirely happy with the docs change, but I don't know how to explain it better without breaking LazyValue out into it's own section.

Closes: #473

ndench committed:
feat(lazy): Add memoized lazy value

ndench committed:
docs update

ndench commented:
Hmm I'm not entirely sure what to do about the CI failures.

ndench commented on src/LazyValue.php:
```diff
@@ -19,27 +19,47 @@ final class LazyValue
+        $this->memoize = $memoize;
+    }
+
+    public static function memoize(callable $factory): self
```
Should we add a named constructor `LazyValue::new` for non-memoized? Would make it consistent, but we can't make the contstructor private for BC.

nikophil replied:
I think this is a good idea, maybe we could add  a depercation layer for using `new Lazy()` directly?
see https://github.com/zenstruck/foundry/blob/1.x/src/FactoryCollection.php#L41-L43 as an example

ndench commented on tests/Unit/FunctionsTest.php:
```diff
@@ -44,6 +47,29 @@ public function faker(): void
     }
 
+    /**
+     * @test
+     */
+    public function lazy(): void
+    {
+        $callback = new LazyValue(fn () => new \stdClass());
+        $value = lazy($callback);
+
+        $this->assertInstanceOf(LazyValue::class, $value);
+        $this->assertNotSame($value(), $value());
+    }
+    /**
+     * @test
+     */
+    public function memoize(): void
+    {
+        $callback = new LazyValue(fn () => new \stdClass());
+        $value = memoize($callback);
+
+        $this->assertInstanceOf(LazyValue::class, $value);
+        $this->assertSame($value(), $value());
+    }
```
These duplicate the LazyValueTest. But I feel like they're necessary to ensure the functions are wired in correctly.

nikophil replied:
that's fine IMO

ndench committed:
self review

nikophil commented on src/LazyValue.php:
```diff
@@ -19,27 +19,47 @@ final class LazyValue
     /** @var callable():mixed */
     private $factory;
 
+    /** @var bool */
+    private $memoize;
+
+    /** @var mixed */
+    private $memoizedValue = null;
```
```suggestion
    private bool $memoize;
    private mixed $memoizedValue = null;
```

nikophil reviewed (changes requested):
nice! :ok_hand: 

> Hmm I'm not entirely sure what to do about the CI failures.

I've fixed it here: https://github.com/zenstruck/foundry/pull/476
once it will be merged, you can rebase `1.x` branch

> I'm not entirely happy with the docs change, but I don't know how to explain it better without breaking LazyValue out into it's own section.

I'd say it's time for `LazyValue` to have its own section. Maybe we can keep the `.. note::` about the lazy, but shorten it and link to the new section?

ndench commented:
Thanks for the feedback @nikophil. I'll make those changes.

ndench committed:
Create LazyValue::new and document LazyValue fully

ndench commented:
@nikophil I've made those those changes. Would love some feedback on my LazyValues documentation section. 

Also, I can't quite figure out how to generate the docs to check the formatting.

kbond commented on docs/index.rst:
```diff
@@ -854,6 +841,53 @@ The following assumes the ``Post`` entity has a many-to-many relationship with `
+~~~~~~~~~~~
+
+The ``getDefaults()`` method is called everytime a factory is instantiated (even if you don't end up
+creating it). Sometimes, you might not want your value calculated every time. For example, ff you have a value for one
```
```suggestion
creating it). Sometimes, you might not want your value calculated every time. For example, if you have a value for one
```

kbond reviewed (commented):
Looks good to me, just a small typo.

> Also, I can't quite figure out how to generate the docs to check the formatting.

You should be able to run `bin/build-docs`, then open `docs/output/index.html`.

ndench committed:
Update docs/index.rst

Co-authored-by: Kevin Bond <kevinbond@gmail.com>

ndench committed:
add Makefile command to generate docs

ndench commented:
Thanks @kbond. I've updated the README and the Makefile for generating the docs (hope that's ok!).

Here's some screenshots of the changes I made to the docs:

## The NOTE underneath **Model factories -> generate**
### Old
![2023-07-04_09-12](https://github.com/zenstruck/foundry/assets/2062388/3ee22f6a-1f5a-4f8c-a74a-9f8f4693a466)
### New
![2023-07-04_09-10](https://github.com/zenstruck/foundry/assets/2062388/48661e91-7678-4578-9626-33f483608213)

## The TIP underneath **Doctrine relationships -> Many-to-one**
### Old
![2023-07-04_09-13](https://github.com/zenstruck/foundry/assets/2062388/2b08b056-0143-4d05-b25d-5702125d3278)
### New
![2023-07-04_09-11](https://github.com/zenstruck/foundry/assets/2062388/fe40e176-7cd2-447a-9c2d-1e3804f56868)

## The new **Lazy values** section
![2023-07-04_10-03](https://github.com/zenstruck/foundry/assets/2062388/131e1c70-5a5e-405b-951d-7ea2454253f8)

kbond commented:
@nikophil, I think this is ready to go but can you check the makefile stuff?

nikophil commented:
Yeah sorry, I'm currently in holidays ⛱️
I might reconnect from next monday 😁

nikophil commented on src/LazyValue.php:
```diff
@@ -19,27 +19,56 @@ final class LazyValue
+
+    public static function new(callable $factory): self
+    {
+        return new self($factory, false);
```
shouldn't we call `self($factory, calledInternally: true)` ?

and `self($factory, memoize: true, calledInternally: true)` in `memoize()`

otherwise, we will trigger our own deprecations

ndench replied:
Oops, good catch!

nikophil commented on tests/Unit/LazyValueTest.php:
I think some tests in this class will trigger deprecations. Either we must use named constructors, or tag these tests with `@group legacy`

ndench replied:
Good point. I'll update them all to use named constructors and add a single legacy test to ensure the constructor still works as expected.

nikophil commented on Makefile:
```diff
@@ -96,6 +96,11 @@ $(PSALM_BIN): vendor bin/tools/psalm/composer.json bin/tools/psalm/composer.lock
 	@touch -c $@ bin/tools/psalm/composer.json bin/tools/psalm/composer.lock
 
+.PHONY: docs
+docs: vendor ### Generate documentation to docs/output
```
:+1:

nikophil reviewed (changes requested):
Hi!

the docs enhancement looks very good to me.

I think a small fix must be done in order to not auto-trigger our own deprecation errors

BTW, you can rebase `1.x` and the tests will be OK

ndench committed:
Correctly handle deprecation internally

ndench committed:
fix tests

ndench committed:
resolve test deprecation and add new memoize test

ndench commented:
Thanks @nikophil. I've rebased and resolved all the internal deprecations. I also added `FactoryTest::lazyMemoized_values_are_only_calculated_once` since there was a LazyValue test in there too.

nikophil commented:
thanks @ndench

nikophil commented:
@ndench I've released this PR in v1.34.0

you'll can soon admire your docs update on symfony.com :wink:

ndench commented:
Thanks @nikophil! Really appreciate your time (and @kbond too)!
